pkg/virt-operator/resource/apply/rbac.go

```diff
@@ -4,183 +4,320 @@ import (
-			r.expectations.RoleBinding.LowerExpectations(r.kvKey, 1, 0)
-			return fmt.Errorf("unable to create rolebinding %+v: %v", rb, err)
+	switch roleType {
+	case TypeRole:
+		role := obj.(*rbacv1.Role)
+
+		updateFunc = func() (err error) {
+			_, err = rbacObj.Roles(namespace).Update(context.Background(), role, metav1.UpdateOptions{})
+			return err
 		}
+	case TypeClusterRole:
+		role := obj.(*rbacv1.ClusterRole)
 
-		log.Log.V(2).Infof("rolebinding %v created", rb.GetName())
-	} else if !objectMatchesVersion(&cachedRb.ObjectMeta, imageTag, imageRegistry, id, r.kv.GetGeneration()) {
-		// Update existing, we don't need to patch for rbac rules.
-		_, err = rbac.RoleBindings(namespace).Update(context.Background(), rb, metav1.UpdateOptions{})
-		if err != nil {
-			return fmt.Errorf("unable to update rolebinding %+v: %v", rb, err)
+		updateFunc = func() (err error) {
+			_, err = rbacObj.ClusterRoles().Update(context.Background(), role, metav1.UpdateOptions{})
+			return err
+		}
+	case TypeRoleBinding:
+		roleBinding := obj.(*rbacv1.RoleBinding)
+
+		updateFunc = func() (err error) {
+			_, err = rbacObj.RoleBindings(namespace).Update(context.Background(), roleBinding, metav1.UpdateOptions{})
+			return err
 		}
+	case TypeClusterRoleBinding:
+		roleBinding := obj.(*rbacv1.ClusterRoleBinding)
 
-		log.Log.V(2).Infof("rolebinding %v updated", rb.GetName())
-	} else {
-		log.Log.V(4).Infof("rolebinding %v already exists", rb.GetName())
+		updateFunc = func() (err error) {
+			_, err = rbacObj.ClusterRoleBindings().Update(context.Background(), roleBinding, metav1.UpdateOptions{})
+			return err
+		}
 	}
 
-	return nil
+	return
 }
 
-func (r *Reconciler) createOrUpdateRole(role *rbacv1.Role,
-	imageTag string,
-	imageRegistry string,
-	id string,
-	namespace string) error {
-
-	if !r.stores.ServiceMonitorEnabled && (role.Name == rbac.MONITOR_SERVICEACCOUNT_NAME) {
-		return nil
+func getRoleTypeName(roleType RoleType) (name string) {
+	switch roleType {
+	case TypeRole:
+		name = "role"
+	case TypeClusterRole:
+		name = "clusterrole"
+	case TypeRoleBinding:
+		name = "rolebinding"
+	case TypeClusterRoleBinding:
+		name = "clusterrolebinding"
 	}
 
-	var err error
-	rbac := r.clientset.RbacV1()
+	return
+}
+
+func getMetaObject(role interface{}, roleType RoleType) (meta *metav1.ObjectMeta) {
+	switch roleType {
+	case TypeRole:
+		role := role.(*rbacv1.Role)
+		meta = &role.ObjectMeta
+	case TypeClusterRole:
+		role := role.(*rbacv1.ClusterRole)
+		meta = &role.ObjectMeta
+	case TypeRoleBinding:
+		roleBinding := role.(*rbacv1.RoleBinding)
+		meta = &roleBinding.ObjectMeta
+	case TypeClusterRoleBinding:
+		roleBinding := role.(*rbacv1.ClusterRoleBinding)
+		meta = &roleBinding.ObjectMeta
+	}
 
-	var cachedR *rbacv1.Role
+	return
+}
 
-	role = role.DeepCopy()
-	obj, exists, _ := r.stores.RoleCache.Get(role)
-	if exists {
-		cachedR = obj.(*rbacv1.Role)
+func enforceAPIGroup(existing interface{}, required interface{}, roleType RoleType) {
```

Comment:
could we instead just pass in  
```
	var existingRoleRef *rbacv1.RoleRef
	var requiredRoleRef *rbacv1.RoleRef
	var existingSubjects []rbacv1.Subject
	var requiredSubjects []rbacv1.Subject
```

since these are the same underlying types no matter the parent object? I think we could pass the ref to the array  and still mutate in place?

Reply:
Do you mean passing them into the function?
This isn't possible because only the "binding" types has `roleRef` and `Subject` attributes. Also, in order to pass those into the function we need to cast the objects inside `createOrUpdate()` function. I think it's best the `createOrUpdate()` will be clean of casting that will only happen inside helper functions.

As I see it, except for casting in parent function, the only thing I can do is to move the loops below to be within the switch-case clauses, but I think it will make the code more duplicated and less clear.

Hope I understood you correctly